state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -1619,33 +1619,45 @@ pub mod pallet {
 						return;
 					};
 
-					let amount_to_refund = swap.input_amount + *remaining_input_amount;
-
-					match &refund_params.refund_destination {
-						AccountOrAddress::ExternalAddress(address) => {
-							Self::egress_for_swap(
-								request.id,
-								amount_to_refund,
-								request.input_asset,
-								address.clone(),
-								None, /* refunds don't use ccm parameters */
-								true, /* refund */
+					let total_input_remaining = swap.input_amount + *remaining_input_amount;
+					let amount_to_refund =
+						match Self::take_refund_fee(total_input_remaining, request.input_asset) {
+							Ok(remaining) => remaining,
+							Err(e) => {
+								log_or_panic!(
+								"Failed to calculate refund fee for swap request {swap_request_id}: {e:?}"
 							);
-						},
-						AccountOrAddress::InternalAccount(account_id) => {
-							Self::deposit_event(Event::<T>::RefundedOnChain {
-								swap_request_id,
-								account_id: account_id.clone(),
-								asset: request.input_asset,
-								amount: amount_to_refund,
-							});
-
-							T::BalanceApi::credit_account(
-								account_id,
-								request.input_asset,
-								amount_to_refund,
-							);
-						},
+								total_input_remaining
+							},
+						};
+
+					if amount_to_refund > 0 {
```

Comment:
We probably should emit an event in the case we're not above 0, else it's hard to see where the flow ended.

Reply:
We have `SwapRequestCompleted` a bit further down - seems like enough?